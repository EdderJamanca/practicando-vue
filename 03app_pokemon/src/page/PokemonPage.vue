<template>
    <h1 v-if="!pokemon">Espere por favor........</h1>
    <div v-else>
        <h1>¿Quien es este pokémon ?</h1>
        <!-- todo: img -->
        <Imagen 
            :pokemonId="pokemon.id" 
            :showpokemon="showPokemon"/>
            
        <Opciones
            :pokemos="pokemonArr"
            @selection-pokemon="checkAnswer"
         />
         
    </div>
    <!-- todo: opciones -->
    <template v-if="showAnswer">
        <h2>{{ message }}</h2>
        <button
            @click="newGame" 
        >
            Nuevo Juego
        </button>
    </template>
</template>


<script>
import Imagen from './components/Imagen.vue';
import Opciones from './components/Opciones.vue';
import getPokemonOptions from './services/getPokemonOptions';

// console.log("hola",getPokemonOptions());
export default {
    components:{
        Opciones,
        Imagen
    },
    data(){
        return { 
            pokemonArr:[],
            pokemon:null,
            showPokemon:null,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        async mixPokemonArray() {
            this.pokemonArr = await getPokemonOptions();
            console.log("this.pokemonArr",this.pokemonArr)
            const rndInt = Math.floor(Math.random()*4);
            // console.log("this.pokemonArr[rndInt]",this.pokemonArr[rndInt])
            this.pokemon = this.pokemonArr[rndInt];
        },
        checkAnswer(selectedId){


            if(selectedId == this.pokemon.id){
                this.message =`correcto, ${this.pokemon.name}`
                this.showPokemon=true;
                this.showAnswer =true;
            }else {
                this.message = `Oops, era ${this.pokemon.name}`
            }
        },
        newGame(){
            this.showPokemon=false;
            this.showAnswer =false;
            this.pokemonArr =[];
            this.pokemon =null;
            this.mixPokemonArray();
        }
    },
    mounted(){
        this.mixPokemonArray();
    }

}
</script>