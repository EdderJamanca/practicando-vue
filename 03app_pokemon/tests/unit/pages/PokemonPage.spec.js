import { shallowMount } from "@vue/test-utils";
// import pokemonPage from "@/page/PokemonPage";

import { pokemons } from "../mocks/pokemons.mock";
import PokemonPage from "@/page/PokemonPage.vue";

describe('Pokemon Component',()=>{
    let wrapper;

    beforeEach(()=>{
        wrapper=shallowMount(PokemonPage);
    })


    test('debe de hacer match con el snapshot',()=>{
        expect(wrapper.html()).toMatchSnapshot();
    })

    test('debe de llamar mixPokemonArray al montar',()=>{

        const mixPokemonArraySpy=jest.spyOn(
            PokemonPage.methods,'mixPokemonArray');

        const wrapper=shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalled();
    });

    test('debe de hacer match con el snapshot cuando cargan los pokemons',()=>{

        const wrapper=shallowMount(PokemonPage,{
            data(){
                return {
                    pokemonArr:pokemons,
                    pokemon:pokemons[0],
                    showPokemon:false,
                    showAnswer:false,
                    message:''
                }
            }
        });

        expect(wrapper.html()).toMatchSnapshot();

    })

    test('debe de mostrar los componentes de Image y option',()=>{

        const wrapper=shallowMount(PokemonPage,{
            data(){
                return {
                    pokemonArr:pokemons,
                    pokemon:pokemons[0],
                    showPokemon:false,
                    showAnswer:false,
                    message:''
                }
            }
        });

        const picture = wrapper.find('imagen-stub')
        const options = wrapper.find('opciones-stub')

        expect( picture.exists() ).toBeTruthy()
        expect( options.exists() ).toBeTruthy()

        
        expect( picture.attributes('pokemonid') ).toBe('5')

        expect( options.attributes('pokemos') ).toBeTruthy()

    })


    test('prueba con checkAnswer',async()=>{

        const wrapper=shallowMount(PokemonPage,{
            data(){
                return {
                    pokemonArr:pokemons,
                    pokemon:pokemons[0],
                    showPokemon:false,
                    showAnswer:false,
                    message:''
                }
            }
        });

        console.log("pokemons[0],",pokemons[0])

        await wrapper.vm.checkAnswer(5);

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showAnswer).toBe(true)

        expect(wrapper.find('h2').text()).toBe(`correcto, ${pokemons[0].name}`)

        await wrapper.vm.checkAnswer(10)

        expect(wrapper.vm.message).toBe( `Oops, era ${pokemons[0].name}`)
    })

})