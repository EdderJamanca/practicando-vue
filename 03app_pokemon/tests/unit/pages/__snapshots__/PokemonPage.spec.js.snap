// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Component debe de hacer match con el snapshot 1`] = `
<h1>Espere por favor........</h1>
<!-- todo: opciones -->
<!--v-if-->
`;

exports[`Pokemon Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quien es este pokémon ?</h1><!-- todo: img -->
  <imagen-stub pokemonid="5" showpokemon="false"></imagen-stub>
  <opciones-stub pokemos="[object Object],[object Object],[object Object],[object Object]"></opciones-stub>
</div>
<!-- todo: opciones -->
<!--v-if-->
`;
